### PR TITLE
ENH: add a function for Gauss-Hermite quadrature

### PR DESCRIPTION
**Pull request by davharris**
[Gauss-Hermite quadrature](https://en.wikipedia.org/wiki/Gauss%E2%80%93Hermite_quadrature) is useful for indefinite intervals involving probability distributions, especially in contexts with infinite mixtures such as generalized linear mixed models. The submitted function is designed to allow users to use this technique without manually shifting and scaling the output of ~~numpy.polynomial.hermite.hermgauss~~ `scipy.special.h_roots`.

This is my first time submitting to a large Python project, and I apologize in advance if there's anything wrong with this PR.  I hope it's useful.

### CODE REVIEW

**Comment by person142** [comment]
Wouldn't it make more sense to add a `method` keyword to `fixed_quad`?

**Comment by davharris** [comment]
Maybe? If they could be combined without cluttering up the API and docs too much, I wouldn't be opposed to combining them, but I'm not sure how that would work.

The main issue I see is the definite versus indefinite distinction.  It's an error to include bounds (`a` and `b`) in Gauss-Hermite quadrature and it's an error to exclude them in Gauss–Legendre quadrature. Meanwhile, `mu_hat` and `sigma_hat` are only used for Gauss-Hermite.

**Comment by davharris** [comment]
Also, it looks like Travis is unhappy that I didn't update the reference guide.  Could someone help me find it? [My searches for the correct file are coming up empty.](https://github.com/scipy/scipy/search?p=1&q=%22refguide%22+OR+%22reference+guide%22&utf8=%E2%9C%93)

Sorry if I'm just being dense.

**Comment by aarchiba** [comment]
If you added this to fixed_quad, there would then become available adaptive and Romberg versions of Gauss-Hermite quadrature. This could be done by simply applying this method when fixed_quad is presented with the interval (-inf,inf), or you could add an optional argument to fixed_quad and others that allowed the specification of which orthogonal polynomials to use - others are useful for half-infinite intervals or integrable singularities or other families of functions.

Also worth considering is that stock GH integration is designed for polynomials times a Gaussian with mean zero and standard deviation one (?). Is it worth providing the option to shift and scale the function of it's centered elsewhere?

**Comment by davharris** [comment]
@aarchiba: 

I don't know what you mean by "adaptive and Romberg versions of Gauss-Hermite quadrature." Could you clarify? 

> Is it worth providing the option to shift and scale the function of it's centered elsewhere?

That's what the `mu_hat` and `sigma_hat` arguments do.  Did you have something else in mind?

**Comment by person142** [comment]
So the signature could be something like `fixed_quad(func, a, b, args, method='gauss-legendre', options=None)`. Parameters like `mu_hat` can be passed through `options`.

**Comment by aarchiba** [comment]
Sorry, I missed mu_hat and sigma_hat.

Currently scipy.integrate provides the functions romberg, quadrature, and fixed_quad to do Gaussian quadrature. The last provides fixed-order integration, quadrature does adaptive integration - increases the order until the result meets some error bound - and romberg uses extrapolation to infinite order to try for higher accuracy. But the other two simply call fixed_quad, so if fixed_quad could handle infinite intervals with GH quadrature, then so could the other two.

More generally, orthopoly1d objects from scipy.special contain roots, weights, and limits, so they could be passed to these functions to allow users to specify alternative sets of orthogonal polynomials. You'd still want something like mu_hat and sigma_hat for the infinite or semi-infinite intervals.

**Comment by person142** [comment]
@davharris: `quadrature` calls `fixed_quad` under the hood.

File: scipy/integrate/quadrature.py [hunk 1]
```diff
@@ -853,3 +860,105 @@ def newton_cotes(rn, equal=0):
...
+    Parameters
+    ----------
+    func : callable
+        A Python function or method to integrate (must accept vector inputs).
+        For best results, `func` should approximately equal a Gaussian
+        density times a low-order polynomial.
+    n : int
+        Order of quadrature integration. This function relies on
+        `numpy.polynomial.hermite.hermgauss`, which has not been tested for
+        order larger than 100.
```
Comment (aarchiba) [inline scipy/integrate/quadrature.py]: scipy.special.h_roots claims to be good up to several thousand.

Comment (davharris) [inline scipy/integrate/quadrature.py]: Thanks. I've swapped it in.

**Comment by davharris** [comment]
Okay, I'd be open to fitting this into `fixed_quad` (and thus `quadrature`) if there's consensus for it. My understanding is that `romb` is not compatible with Gauss-Hermite because it requires equally-spaced samples.

What do other folks think?

**Comment by ev-br** [comment]
A big +1 from me for adding a choice of quadrature rules to `quadrature`/`fixed_quad`. (I haven't looked at the details)

File: scipy/integrate/quadrature.py [hunk 2]
```diff
@@ -853,3 +859,105 @@ def newton_cotes(rn, equal=0):
...
+        nodes (e.g. the standard deviation if `func` is proportional to a
+        probability density).
+    zero_nan : bool, optional
+        If `func(x)` returns `nan` (e.g. if a quadrature point falls outside the
+        domain of x), should this value be coerced to zero? Default is False.
+
+    Returns
+    -------
+    val : float
+        Gauss-Hermite quadrature approximation to the indefinite integral of
```
Comment (ev-br) [inline scipy/integrate/quadrature.py]: Indefinite integral? I thought that's a synonym for antiderivative, which is not what this function computes.

Comment (davharris) [inline scipy/integrate/quadrature.py]: Looks like you're right. Maybe the term I was looking for is "improper integral"? http://mathworld.wolfram.com/ImproperIntegral.html

Or maybe it's better just to say that it integrates from `-Inf` to `Inf`.

**Comment by aarchiba** [comment]
> So the signature could be something like `fixed_quad(func, a, b, args, method='gauss-legendre', options=None)`. Parameters like `mu_hat` can be passed through `options`.

I would prefer a signature of `fixed_quad(func, a, b, n, args=None, orthopoly=None, mu_hat=None, sigma_hat=None)`. Then if no specific orthogonal polynomial is provided, the interval (a,b) is used to choose: if finite, use the current default, if (-inf,inf) use Gauss-Hermite, if half-infinite perhaps Laguerre polynomials? `mu_hat` and `sigma_hat` would be ignored (raising appropriate errors) if the interval was not infinite.  If `orthopoly` is provided, then it should be an `orthopoly1d` object (of degree equal to n); it contains a `limits` attribute that can be used to rescale the x axis as needed. 

Adaptive and Romberg quadrature will need to be passed a callable capable of generating orthopoly1d objects for various n; perhaps this function should get the same if a different set of orthogonal polynomials is wanted.

**Comment by charris** [comment]
@aarchiba Unfortunately, orthopoly1d subclasses `poly1d` and is initialized by the roots of the polynomial, so it is not numerically stable. If we stick to monic polynomials defined by the roots, a simple product `(x - r0)*...*(x - rn)` would be better. A normalization factor could be added to manage the leading coefficient of the polynomial and could be easily computed from the recursion or even be made part of the product term by term.

**Comment by aarchiba** [comment]
@charris actually you can read off the roots and weights - which are what we would use here - and also they provide evaluation through the appropriate special function. Only if you do arithmetic with them do you find yourself stuck with coefficients.

**Comment by charris** [comment]
@aarchiba Oh, right. Nevermind ;)

**Comment by davharris** [comment]
It's really cool to see all the discussion this has raised.  Here are some thoughts:

### With regard to `orthopoly`:

I don't know enough about orthogonal polynomials to implement this part by myself, but I'd be happy to do the code where `orthopoly=False` or to work with someone that understands them better.

### With regard to the function signature:

* I don't think there's enough information in `a` and `b` to infer the polynomial, if we include the [full menu of polynomial types](https://en.wikipedia.org/wiki/Gaussian_quadrature#Other_forms). In particular, I don't think the code could distinguish among Legendre, Jacobi, and the two Chebyshev ones.

* I wish `a` and `b` weren't so early in the function signature. It'll be annoying to skip over them every time to get to `n` and other downstream arguments.  Probably no good solution that wouldn't break backward compatibility or add new functions with duplicate functionality, though.

### With regard to Romberg integration:

Romberg integration is separate, isn't it? I've never used Romberg's method before, so I could be wrong. But I don't see any calls to `fixed_quad` in the `romb` code and Wikipedia says that Romberg's method uses equally-spaced points instead of points from orthogonal polynomials.

**Comment by aarchiba** [comment]
> With regard to orthopoly:

> I don't know enough about orthogonal polynomials to implement this part by myself, but I'd be happy to do the code where orthopoly=False or to work with someone that understands them better.

Actually, you just retrieve the roots and weights from the orthopoly object instead of calling scipy.special.h_roots. (It's constructed in such a way that it calls the appropriate special functions.) So it's really not very different, and the existing code would simply happen by defaulting the value of orthopoly. (And you need to pull the limits out of the orthopoly object to know how to scale the input interval.) But I'm willing to work on the more-general code.

> With regard to the function signature:

> I don't think there's enough information in a and b to infer the polynomial, if we include the full menu of polynomial types. In particular, I don't think the code could distinguish among Legendre, Jacobi, and the two Chebyshev ones.

> I wish a and b weren't so early in the function signature. It'll be annoying to skip over them every time to get to n and other downstream arguments. Probably no good solution that wouldn't break backward compatibility or add new functions with duplicate functionality, though.

You certainly can't pick from the full set. But there are three cases of (a,b): finite, half-infinite, and infinite. We can pick a sensible default for each one:  the existing Legendre, perhaps Laguerre, and Gauss-Hermite respectively. If people care about the weight functions, they can specify the orthopoly argument and get any set scipy supports, or any set they are willing to implement.

> With regard to Romberg integration:

> Romberg integration is separate, isn't it? I've never used Romberg's method before, so I could be wrong. But I don't see any calls to fixed_quad in the romb code and Wikipedia says that Romberg's method uses equally-spaced points instead of points from orthogonal polynomials.

You're quite right. One *could* do Romberg integration with orthogonal polynomials, but you can't (normally) re-use evaluation points, and anyway scipy doesn't. Sorry, I misread the code.

**Comment by person142** [comment]
Once you've constructed an instance of `orthopoly1d` the quadrature is quite simple though, e.g.

``` python
>>> p = special.legendre(5)
>>> (f(p.roots)*p.weights[:,1]).sum()
```
Why have a function that reproduces the one line? I like `fixed_quad` because a user doesn't have to know that Gaussian quadrature has anything to do with orthogonal polynomials; by the time they do they're probably fine with reproducing the above.

**Comment by person142** [comment]
Also @davharris: re

> I wish a and b weren't so early in the function signature. It'll be annoying to skip over them every time to get to n and other downstream arguments. Probably no good solution that wouldn't break backward compatibility or add new functions with duplicate functionality, though.

it would be possible to default to `a=None, b=None`; as long as the arguments aren't permuted that is backwards-compatible.

**Comment by aarchiba** [comment]
@person142 once you add the appropriate scaling for alternative intervals (or means and standard deviations), and adaptive quadrature, it becomes rather more than one line. But yes, it should be usable without knowing much about orthogonal polynomials.

File: scipy/integrate/quadrature.py (hunk 2 above)
Comment (mdhaber) [reply scipy/integrate/quadrature.py:897]: Yes, "improper" is the proper term : ) Mentioning the infinite limits of integration is good, too.

**Comment by mdhaber** [comment]
Superseded by gh-18650.

**Pull request closed without merging**